taskvine/src/worker/vine_cache.c

```diff
@@ -644,6 +647,17 @@ vine_cache_status_t vine_cache_ensure(struct vine_cache *c, const char *cachenam
 		f->process = p;
 	}
 
+	while (list_size(c->current_transfers) > 5) {
```

Comment:
Returning pending instead of polling here would imply that we already forked all transfer processes and they have the knowledge to block themselves until the current amount of transfers is below the limit. 

Having the transfer processes synchronize was something I investigated but you were rightfully wary of semaphores, and I started to need a number of shared memory items which I figured was discouraged.

if the transfer processes must remain simple in their nature, then we will need to poll somewhere on the main process to limit the transfers

Reply:
Right, but look at the main loop in `vine_worker_serve_manager`.  It's already the case that the main program polls a variety of things and then goes back to sleep until something changes.  It looks at `vine_cache_wait` to see if any transfers have complete and also `vine_process_ready_to_run_now->vine_sandbox_ensure->vine_cache_ensure` to see if a task is satisfied.  This is just one more way for the cache to say "not yet" when the main program asks if a particular file is ready.

Perhaps `ensure` is a misleading name and this function would be better called `vine_cache_isready` or similar.